clients/src/main/java/org/apache/kafka/common/header/internals/RecordHeaders.java

```diff
@@ -0,0 +1,214 @@
+    
+    private final List<Header> headers;
+    private volatile boolean isClosed = false;
+
+    public RecordHeaders() {
+        this((Iterable<Header>) null);
+    }
+
+    public RecordHeaders(Header[] headers) {
+        if (headers == null) {
+            this.headers = new ArrayList<>();
+        } else {
+            //new ArrayList using efficient Arrays.copyOf the given the Arrays.asList wrapper.
+            this.headers = new ArrayList<>(Arrays.asList(headers));
+        }
+    }
+    
+    public RecordHeaders(Iterable<Header> headers) {
+        if (headers == null) {
+            //new ArrayList, empty as no headers
+            this.headers = new ArrayList<>();
+        } else if (headers instanceof RecordHeaders) {
+            //new ArrayList using efficient Arrays.copyOf the backing list in given RecordHeaders
+            this.headers = new ArrayList<>(((RecordHeaders) headers).headers);
+        } else if (headers instanceof Collection) {
+            //new ArrayList using efficient Arrays.copyOf the given collection.
+            this.headers = new ArrayList<>((Collection<Header>) headers);
+        } else {
+            //new ArrayList iterating through to copy the headers.
+            this.headers = new ArrayList<>();
+            Iterator<Header> iterator = headers.iterator();
+            while (iterator.hasNext()) {
+                this.headers.add(iterator.next());
+            }
+        }
+    }
+
+    @Override
+    public Headers add(Header header) throws IllegalStateException {
+        checkClosed();
+        headers.add(header);
+        return this;
+    }
+
+    @Override
+    public Headers add(String key, byte[] value) throws IllegalStateException {
+        return add(new RecordHeader(key, value));
+    }
+
+    @Override
+    public Headers remove(String key) throws IllegalStateException {
+        checkClosed();
+        checkKey(key);
+        Iterator<Header> iterator = iterator();
+        while (iterator.hasNext()) {
+            if (iterator.next().key().equals(key)) {
+                iterator.remove();
+            }
+        }
+        return this;
+    }
+
+    @Override
+    public Header lastHeader(String key) {
+        checkKey(key);
+        for (int i = headers.size() - 1; i >= 0; i--) {
+            Header header = headers.get(i);
+            if (header.key().equals(key)) {
+                return header;
+            }
+        }
+        return null;
+    }
+
+    @Override
+    public Iterable<Header> headers(final String key) {
+        checkKey(key);
+        return new Iterable<Header>() {
+            @Override
+            public Iterator<Header> iterator() {
+                return new FilterByKeyIterator(headers.iterator(), key);
+            }
+        };
+    }
+
+    @Override
+    public Iterator<Header> iterator() {
+        return closeAware(headers.iterator());
+    }
+
+    @Override
+    public void close() throws IOException {
```

Comment:
I was thinking about this and maybe `close` is not the right name. Because we can still use the class, we simply cannot mutate it any more. The name that came to mind is `seal`, but maybe that's not clear either. We could do the boring `closeForUpdates` or something like that. Thoughts?